src/main/scala/ch/epfl/bluebrain/nexus/iam/acls/Acls.scala

```diff
@@ -106,49 +109,59 @@ class Acls[F[_]](agg: Agg[F], index: AclsIndex[F])(implicit F: Monad[F],
     index.get(path, ancestors, self)(caller.identities)
 
-  private def check(path: Path, fetched: F[OptResourceAccessControlList], self: Boolean)(
-      implicit caller: Caller): F[OptResourceAccessControlList] = {
-    fetched flatMap {
-      case acls if self                            => F.pure(acls.map(_.map(_.filter(caller.identities))))
-      case Some(acls) if containsWrite(acls.value) => F.pure(Some(acls))
-      case acls =>
-        ancestorsContainsWrite(path.parent).map {
-          case true  => acls
-          case false => acls.map(_.map(_.filter(caller.identities)))
-        }
-    }
-  }
-
-  private def fetchUnsafe(path: Path): F[OptResourceAccessControlList] =
+  private def fetchUnsafe(path: Path): F[ResourceOpt] =
     agg.currentState(path.asString).map(stateToAcl(path, _))
 
-  private def fetchUnsafe(path: Path, rev: Long): F[OptResourceAccessControlList] =
+  private def fetchUnsafe(path: Path, rev: Long): F[ResourceOpt] =
     agg
       .foldLeft[AclState](path.asString, Initial) {
         case (state, event) if event.rev <= rev => next(state, event)
         case (state, _)                         => state
       }
       .map(stateToAcl(path, _))
 
-  private def stateToAcl(path: Path, state: AclState): OptResourceAccessControlList =
+  private def stateToAcl(path: Path, state: AclState): ResourceOpt =
     (state, path) match {
       case (Initial, initAcl.path) => Some(initAcl.acl)
       case (Initial, _)            => None
-      case (c: Current, _)         => Some(c.toResource)
+      case (c: Current, _)         => Some(c.resource)
     }
 
-  private def containsWrite(acl: AccessControlList)(implicit caller: Caller): Boolean =
-    acl.value.exists { case (ident, perms) => caller.identities.contains(ident) && perms.contains(writeAcls) }
+  private def check(path: Path, permission: Permission)(implicit caller: Caller): F[Unit] =
+    hasPermission(path, permission, ancestors = true)
+      .ifM(F.unit, F.raiseError(AccessDenied(path.toIri, permission)))
 
-  private def ancestorsContainsWrite(path: Path)(implicit caller: Caller): F[Boolean] =
-    if (path.isEmpty)
-      F.pure(false)
+  private def filterSelf(opt: ResourceOpt)(implicit caller: Caller): ResourceOpt =
+    opt.map(_.map(acl => acl.filter(caller.identities)))
+
+  def hasPermission(path: Path, perm: Permission, ancestors: Boolean = true)(implicit caller: Caller): F[Boolean] = {
+    def hasPermission(p: Path): F[Boolean] =
+      fetchUnsafe(p).map {
+        case Some(res) => res.value.hasPermission(caller.identities, perm)
+        case None      => false
+      }
+    if (!ancestors) hasPermission(path)
     else
-      fetchUnsafe(path).flatMap {
-        case Some(acls) if containsWrite(acls.value) => F.pure(true)
-        case _ if path == Path./                     => F.pure(false)
-        case _                                       => ancestorsContainsWrite(path.parent)
+      ancestorsOf(path).foldLeftM(false) {
```

Comment:
why do we need that? 
Why is not enough with
```scala
def hasPermission(p: Path): F[Boolean] =
      fetchUnsafe(p).flatMap {
        case Some(res) if res.value.hasPermission(caller.identities, perm) => F.pure(true)
        case _ if p == Path./=> F.pure(false)
        case _ if ancestors => hasPermission(p.parent)
        case _ => F.pure(false)
      }
```
We don't need the `ancestorsOf` method.